### DISCUSSION TITLE
Need information on customising webusb_serial example

### DISCUSSION THREAD

**Discussion by vinayb-imdt** [hathach/tinyusb#1465] [General]
Hello,

I have the tinyUSB ported to a custom board that has the same USB controller (synopsys) as in SM32F723E discovery board. I had the cdc_msc example working and I'm now working on the webusb_serial example. My goal is to create a Vendor specific device class that enumerates on Windows 10 as a winusb device. My plan is to remove the CDC part from the configuration descriptor of this example. Then remove the webusb part from the BOS descriptor of this example and only retain the winusb part. Then use the read and write functions of the Vendor class to communicate with the device using a Windows application through WinUSB interface.

I'm currently facing these two issue and need help to solve them:

1. In the WebUSB_Serial example as is, whatever has been typed in the "sender" section of the "example.tinyusb.org/webusb-serial/" URL is received in the serial console , but characters typed in the serial console are not seen in the "receiver" section of the WebUSB URL. Characters typed on the "sender" box are also not received in the "receiver" box.  I did check that the function "tud_vendor_n_write()" is getting called, I also commented the line "if (tu_fifo_count(&p_itf->tx_ff) >= CFG_TUD_VENDOR_EPSIZE)" so that meybe_transmit() is called explicitly but the data is not received. Any pointers on what could be the issue?
2. If I remove the CDC part from the config descriptor, Windows reports the device in device manager but shows "invalid configuration descriptor" . Is it not possible to remove the CDC part from the configuration descriptor and have the device enumerate as only WinUSB device?

Please advice.

Thanks

**Comment by vinayb-imdt** [comment]
OK solved the "invalid configuration descriptor" issue, but still the device fails to enumerate as WinUSB device and WinUSB.sys is not loaded. Below are my changes:
`
//--------------------------------------------------------------------+
// Device Descriptors
//--------------------------------------------------------------------+
tusb_desc_device_t const desc_device =
{
    .bLength            = sizeof(tusb_desc_device_t),
    .bDescriptorType    = TUSB_DESC_DEVICE,
    .bcdUSB             = 0x0210, // at least 2.1 or 3.x for BOS & webUSB

    // Use Interface Association Descriptor (IAD) for CDC
    // As required by USB Specs IAD's subclass must be common class (2) and protocol must be IAD (1)
    //.bDeviceClass       = TUSB_CLASS_MISC,
    //.bDeviceSubClass    = MISC_SUBCLASS_COMMON,
    //.bDeviceProtocol    = MISC_PROTOCOL_IAD,
    .bDeviceClass       = TUSB_CLASS_VENDOR_SPECIFIC,
    .bDeviceSubClass    = 0x00,
    .bDeviceProtocol    = 0x00,
    .bMaxPacketSize0    = CFG_TUD_ENDPOINT0_SIZE,

    .idVendor           = 0xCafe,
    .idProduct          = USB_PID,
    .bcdDevice          = 0x0100,

    .iManufacturer      = 0x01,
    .iProduct           = 0x02,
    .iSerialNumber      = 0x03,

    .bNumConfigurations = 0x01
};
//--------------------------------------------------------------------+
// Configuration Descriptor
//--------------------------------------------------------------------+
enum
{
  //ITF_NUM_CDC = 0,
  //ITF_NUM_CDC_DATA,
  ITF_NUM_VENDOR = 0,
  ITF_NUM_TOTAL
};

//#define CONFIG_TOTAL_LEN    (TUD_CONFIG_DESC_LEN + TUD_CDC_DESC_LEN + TUD_VENDOR_DESC_LEN)
#define CONFIG_TOTAL_LEN    (TUD_CONFIG_DESC_LEN + TUD_VENDOR_DESC_LEN)
uint8_t const desc_configuration[] =
{
  // Config number, interface count, string index, total length, attribute, power in mA
  TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN, 0x00, 100),

  // Interface number, string index, EP notification address and size, EP data address (out, in) and size.
  //TUD_CDC_DESCRIPTOR(ITF_NUM_CDC, 4, 0x81, 8, EPNUM_CDC_OUT, 0x80 | EPNUM_CDC_IN, TUD_OPT_HIGH_SPEED ? 512 : 64),

  // Interface number, string index, EP Out & IN address, EP size
  TUD_VENDOR_DESCRIPTOR(ITF_NUM_VENDOR, 5, EPNUM_VENDOR_OUT, 0x80 | EPNUM_VENDOR_IN, TUD_OPT_HIGH_SPEED ? 512 : 64)
};
`
I can see that the tud_descriptor_bos_cb() gets called and I do get the request for MS OS 2.0 descriptor. But unable to figure out why the enumerations fails to be identified as a WinUSB device.

Thanks,

**Comment by hathach** [comment]
are you runnning webusb_serial example as it is (un-modified) with makefile. If not then, please try with makefile before doing any ports to prevent similar issue regarding porting with cdc_msc example previously.

**Reply by vinayb-imdt** [reply]
@hathach , I have now tried the WebUSB_Serial example as it is with makefile on the STM32F723E discovery board which is one of the supported boards. The "receiver" box is unable to receive what  has been typed in the "Sender" box, the USB Serial COM port that gets created from this example does receive the characters typed in "Sender" box. Not sure if this issue is related to the Vendor class write function or the WebUSB API. Below is the snapshot of the test.

![image](https://user-images.githubusercontent.com/92086407/169157903-dd11104f-8ca6-4029-82df-d9b49d0418ea.png)

**Comment by vinayb-imdt** [comment]
I eventually have managed to remove the CDC from the WebUSB_Serial example application and modified it to only have Vendor class device,  coming up as a WinUSB device with read and write working. I had to change the endpoints used for vendor class interface. This is all working good on the custom platform. However I couldn't get the WebUSB example to work on STM32F723E discovery board even after changing the Vendor device interface endpoints. Below is what I had to do get the Vendor class come up as WinUSB device on the custom board.

1. Modified tud_descriptor_string_cb() to return Microsoft OS compatibility string descriptor for string descriptor request at index 0xEE. This is to inform MS windows that it is a WCID complaint device. 
2. Next Windows will request vendor specific request using the Vendor_Code which we specified in the  Microsoft OS compatibility string descriptor. We have to now return extended compatibility ID descriptor. Windows now has information to load winusb.sys for our custom device.
3. I also changed the .bcdUSB to 0x0200 in device descriptor, so that the Windows host will not request BOS descriptor.

to get the write functionality working, I had to synchronising the application to wait for DATA-In to signal that previous write was successfully completed. I did this by defining the tud_vendor_tx_cb() function and used a binary semaphore to synchronise with the tud_vendor_n_write() function.

Thanks,

**Reply by hathach** [reply]
string index of 0xEE is the old way i.e MS OS Desc v1, the v2 use BOS descriptors which is available with USB specs 2.1. Anyway, I am glad you get it working. V1 may work with more windows version than v2.
